[Clang] [C23] Implement N2653: u8 strings are char8_t[]

MitalAshok opened the pull request:
https://www.open-std.org/jtc1/sc22/wg14/www/docs/n2653.htm

Closes #97202

MitalAshok committed:
[Clang] [C23] Implement N2653: u8 strings are char8_t[]

MitalAshok commented:
CC @AaronBallman

Also the clang-format issues are intentional to fit the style of the surrounding lines

AaronBallman commented on clang/lib/Frontend/InitPreprocessor.cpp:
```diff
@@ -1342,8 +1342,10 @@ static void InitializePredefinedMacros(const TargetInfo &TI,
-      DEFINE_LOCK_FREE_MACRO(CHAR8_T, Char); // Treat char8_t like char.
+    // char8_t has the same representation / width as unsigned
+    // char in C++ and is a typedef for unsigned char in C23
+    if (LangOpts.Char8 || LangOpts.C23)
```
Would it make more sense for `Char8` to be set by default based on `C23`?

MitalAshok replied:
Looks like that also controls if `char8_t` is a keyword/builtin type, which it isn't in C23. (`LangOpts.Char8` is also false in C++20 with `-fno-char8_t`)

AaronBallman replied:
Ah! Good call, thank you!

MitalAshok committed:
Char array initialized from u8 string was fixed in C++20 as a DR

https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2022/p2513r4.html

MitalAshok committed:
Define ATOMIC_CHAR8_T_LOCK_FREE only when available

This should be equivalent to:

\#if (defined(__STDC_VERSION__) && __STDC_VERSION__ > 202311L) || \
    (defined(__cpp_char8_t) && __cpp_char8_t >= 201811L)

MitalAshok commented on clang/lib/Headers/stdatomic.h:
```diff
@@ -104,6 +107,7 @@ typedef _Atomic(long)               atomic_long;
 typedef _Atomic(unsigned long)      atomic_ulong;
 typedef _Atomic(long long)          atomic_llong;
 typedef _Atomic(unsigned long long) atomic_ullong;
+typedef _Atomic(unsigned char)      atomic_char8_t;
 typedef _Atomic(uint_least16_t)     atomic_char16_t;
 typedef _Atomic(uint_least32_t)     atomic_char32_t;
```
Should the definition of `atomic_char8_t` also be gated behind C23/C++20?

Also in C++20, this should be `_Atomic(char8_t)`, but the next two are also wrong in C++11 (`_Atomic(char16_t)`/`_Atomic(char32_t)`)

AaronBallman replied:
Let's address that one in a separate PR, but yes `atomic_char8_t` should be gated behind C23 because the identifier isn't reserved until C23. For C++, it should be gated behind C++20.

As for the definition, technically in C they should be using `char8_t` and friends as well, but those are defined in `uchar.h` which we don't vend with the compiler (should we be vending that one in freestanding?). GCC seems to use `unsigned char` directly rather than use a transitive include, so I think that approach is fine for us to take as well.

MitalAshok replied:
It seems like this header doesn't properly implement [P0943R6](https://wg21.link/P0943R6) in C++23 (Where `#define _Atomic(T) std::atomic<T>` and all these typedefs should just be `using std::atomic_*`.

So `::atomic_char8_t` should be `std::atomic<char8_t>`, not `std::atomic<unsigned char>` (nor `unsigned char _Atomic`), but that is outside the scope of this PR. I'll just make this match the surrounding typedefs for now.

MitalAshok committed:
Merge branch 'main' into u8_uchar

MitalAshok committed:
[Headers] Gate atomic_char8_t behind C23/C++20[-fchar8_t]

MitalAshok committed:
Add __CHAR8_TYPE__ predefined macro to be compatible with GCC

MitalAshok commented on clang/lib/Frontend/InitPreprocessor.cpp:
```diff
@@ -1165,6 +1165,8 @@ static void InitializePredefinedMacros(const TargetInfo &TI,
   DefineType("__WCHAR_TYPE__", TI.getWCharType(), Builder);
   DefineType("__WINT_TYPE__", TI.getWIntType(), Builder);
   DefineTypeSizeAndWidth("__SIG_ATOMIC", TI.getSigAtomicType(), TI, Builder);
+  if (LangOpts.Char8 || LangOpts.C23)
+    DefineType("__CHAR8_TYPE__", TI.UnsignedChar, Builder);
```
This matches GCC behaviour: https://godbolt.org/z/6Eax3eqrd

But I'm not sure why it's defined in C++ at all

AaronBallman replied:
Hmmm that is surprising, because it expands to `unsigned char` and not `char8_t` there: https://godbolt.org/z/Y96bPbj8e and this seems to have changed between GCC 8 and GCC 9: https://godbolt.org/z/EaEPKr59E

It may be worth opening a GCC bug to see if this was intentional or not.

MitalAshok replied:
According to https://gcc.gnu.org/legacy-ml/gcc-patches/2018-11/msg00295.html `__CHAR8_TYPE__` is "the underlying type of `char8_t`" in C++. `char8_t` in C++ was added in GCC9 which explains the difference between GCC8 and 9

And this made me realise `-std=c++11 -fchar8_t` sets `LangOpts.Char8` in C++11, which I think this patch handles fine (See also: #97601).
Also `-std=c23 -fchar8_t` is a mode that exists. It makes `typeof(u8'x')` `char8_t` and `typeof(u8"")` `char8_t[1]`, even with this patch, which I guess is fine. Pending #55373 if `-std=c23 -fchar8_t` even makes sense.

cor3ntin replied:
I'm certainly concerned with the impact of that in C++. It should not be defined in C++ (ie, char9_t is not unsigned char in C++)

AaronBallman replied:
FWIW, I think https://github.com/llvm/llvm-project/issues/55373 should be fixed so that the option is rejected in C modes, but that's somewhat orthogonal to this patch.

MitalAshok committed:
Tentatively remove macros/atomic_char8_t typedef in C++

AaronBallman commented on clang/test/C/C2x/n2653.c:
```diff
@@ -0,0 +1,29 @@
+// RUN: %clang_cc1 -verify=c23 -std=c23 %s
+// RUN: %clang_cc1 -verify=c17 -std=c17 %s
```
You need to add `-ffreestanding` here otherwise your include below is using whatever it finds on the system rather than our header.

MitalAshok replied:
That still seems to fail on the CI:

```
******************** TEST 'Clang :: C/C23/n2653.c' FAILED ********************
Exit Code: 1

Command Output (stderr):
--
RUN: at line 1: /var/lib/buildkite-agent/builds/linux-56-59b8f5d88-vhr99-1/llvm-project/github-pull-requests/build/bin/clang -cc1 -internal-isystem /var/lib/buildkite-agent/builds/linux-56-59b8f5d88-vhr99-1/llvm-project/github-pull-requests/build/lib/clang/19/include -nostdsysteminc -ffreestanding -verify=c23 -std=c23 /var/lib/buildkite-agent/builds/linux-56-59b8f5d88-vhr99-1/llvm-project/github-pull-requests/clang/test/C/C23/n2653.c
+ /var/lib/buildkite-agent/builds/linux-56-59b8f5d88-vhr99-1/llvm-project/github-pull-requests/build/bin/clang -cc1 -internal-isystem /var/lib/buildkite-agent/builds/linux-56-59b8f5d88-vhr99-1/llvm-project/github-pull-requests/build/lib/clang/19/include -nostdsysteminc -ffreestanding -verify=c23 -std=c23 /var/lib/buildkite-agent/builds/linux-56-59b8f5d88-vhr99-1/llvm-project/github-pull-requests/clang/test/C/C23/n2653.c
error: 'c23-error' diagnostics seen but not expected: 
  File /var/lib/buildkite-agent/builds/linux-56-59b8f5d88-vhr99-1/llvm-project/github-pull-requests/clang/test/C/C23/n2653.c Line 11: missing
1 error generated.

--

********************
```

So I've just removed the test for <stdatomic.h> (`-nostdinc` seems to disable the clang header too)

AaronBallman replied:
I don't think we should remove the test for stdatomic,h, that's an important property for us to test and may be papering over a bug here. The only situation we should be getting the system header is:
```
#if __STDC_HOSTED__ &&                                                         \
    __has_include_next(<stdatomic.h>) &&                                       \
    (!defined(_MSC_VER) || (defined(__cplusplus) && __cplusplus >= 202002L))
# include_next <stdatomic.h>
#else
...
so `-ffreestanding` really should only include our header. What do you get from `-E` on that machine? Are we actually including the system header anyway, or did we mess up our header somehow?
```

MitalAshok replied:
I think I ran the wrong tests then misread some `-E` output... Should be fixed now (`__STDC_VERSION__ > 202311L` -> `__STDC_VERSION__ >= 202311L`)

AaronBallman reviewed (commented):
Precommit CI has found relevant failures:
```
_bk;t=1720277739321********************

_bk;t=1720277739321Failed Tests (1):

_bk;t=1720277739321  Clang :: C/C2x/n2653.c

_bk;t=1720277739321

_bk;t=1720277739321

_bk;t=1720277739321Testing Time: 261.44s
```

MitalAshok committed:
Merge branch 'main' into u8_uchar

MitalAshok committed:
Use Clang header in test

MitalAshok committed:
Remove test for <stdatomic.h> in C23/n2653.c

Still fails with -nostdsysteminc -ffreestanding (seems to be pulling in other system headers still)

MitalAshok committed:
Oops

AaronBallman reviewed (approved):
LGTM, but please wait for @cor3ntin to give a final sign-off before landing.

cor3ntin reviewed (approved):
LGTM!

cor3ntin committed:
Merge branch 'main' into u8_uchar
